include/oneapi/dpl/internal/random_impl/uniform_real_distribution.h

```diff
@@ -128,210 +128,148 @@ class uniform_real_distribution
-        __res = ((__res - static_cast<scalar_type>(__engine.min())) /
-                 (1 + static_cast<scalar_type>(__engine.max() - __engine.min()))) *
-                    (__params.second - __params.first) +
-                __params.first;
-        return __res;
+        return scale_and_displace(static_cast<scalar_type>(__engine()), __params, __engine);
     }
 
-    template <int _Ndistr, int _Negnine, class _Engine>
-    typename ::std::enable_if<((_Ndistr < _Negnine) & (_Ndistr != 0)), result_type>::type
+    template <int _Ndistr, int _Nengine, class _Engine>
+    typename ::std::enable_if<((_Ndistr < _Nengine) & (_Ndistr != 0)), result_type>::type
     generate(_Engine& __engine, const param_type& __params)
     {
         auto __engine_output = __engine(_Ndistr);
         result_type __res;
         for (int __i = 0; __i < _Ndistr; ++__i)
-        {
-            __res[__i] = static_cast<scalar_type>(__engine_output[__i]);
-            __res[__i] =
-                ((__res[__i] - __engine.min()) / (1 + static_cast<scalar_type>(__engine.max() - __engine.min()))) *
-                    (__params.second - __params.first) +
-                __params.first;
-        }
-
+            __res[__i] = scale_and_displace(static_cast<scalar_type>(__engine_output[__i]), __params, __engine);
         return __res;
     }
 
-    template <int _Ndistr, int _Negnine, class _Engine>
-    typename ::std::enable_if<((_Ndistr < _Negnine) & (_Ndistr == 0)), result_type>::type
+    template <int _Ndistr, int _Nengine, class _Engine>
+    typename ::std::enable_if<((_Ndistr < _Nengine) & (_Ndistr == 0)), result_type>::type
     generate(_Engine& __engine, const param_type& __params)
     {
-        scalar_type __res = static_cast<scalar_type>(__engine(1)[0]);
-        __res = ((__res - __engine.min()) / (1 + static_cast<scalar_type>(__engine.max() - __engine.min()))) *
-                    (__params.second - __params.first) +
-                __params.first;
-        return __res;
+        return scale_and_displace(static_cast<scalar_type>(__engine(1)[0]), __params, __engine);
     }
 
-    template <int _Ndistr, int _Negnine, class _Engine>
-    typename ::std::enable_if<((_Ndistr > _Negnine) & (_Negnine != 0)), result_type>::type
+    template <int _Ndistr, int _Nengine, class _Engine>
+    typename ::std::enable_if<((_Ndistr > _Nengine) & (_Nengine != 0)), result_type>::type
     generate(_Engine& __engine, const param_type& __params)
     {
         sycl::vec<scalar_type, _Ndistr> __res;
         int __i;
-        constexpr int __tail_size = _Ndistr % _Negnine;
-        for (__i = 0; __i < _Ndistr - __tail_size; __i += _Negnine)
+        constexpr int __tail_size = _Ndistr % _Nengine;
+        for (__i = 0; __i < _Ndistr - __tail_size; __i += _Nengine)
         {
             auto __engine_output = __engine();
             auto __res_tmp = __engine_output.template convert<scalar_type, sycl::rounding_mode::rte>();
```

Comment:
Should these 2 lines be within the loop? Is there any dependency on `__i`?

Reply:
Yes, line 196 has __res[__i + __j]